Test failure with coordinates: 'NoneType' object has no attribute 'to_geodetic'

- [ ] Revert #11319

There is a test failure in `coordinates` in the 32bits job:
` 'NoneType' object has no attribute 'to_geodetic'`
https://github.com/astropy/astropy/pull/11274/checks?check_run_id=1772534200
```
_______________ test_skyoffset[inradec0-expectedlatlon0-tolsep0] _______________

inradec = <Quantity [45., 45.] deg>, expectedlatlon = <Quantity [0., 0.] deg>
tolsep = <Quantity 0.001 arcsec>, originradec = <Quantity [45., 45.] deg>

    @pytest.mark.parametrize("inradec,expectedlatlon, tolsep", [
        ((45, 45)*u.deg, (0, 0)*u.deg, .001*u.arcsec),
        ((45, 0)*u.deg, (0, -45)*u.deg, .001*u.arcsec),
        ((45, 90)*u.deg, (0, 45)*u.deg, .001*u.arcsec),
        ((46, 45)*u.deg, (1*np.cos(45*u.deg), 0)*u.deg, 16*u.arcsec),
        ])
    def test_skyoffset(inradec, expectedlatlon, tolsep, originradec=(45, 45)*u.deg):
        origin = ICRS(*originradec)
        skyoffset_frame = SkyOffsetFrame(origin=origin)
    
        skycoord = SkyCoord(*inradec, frame=ICRS)
        skycoord_inaf = skycoord.transform_to(skyoffset_frame)
        assert hasattr(skycoord_inaf, 'lon')
        assert hasattr(skycoord_inaf, 'lat')
        expected = SkyCoord(*expectedlatlon, frame=skyoffset_frame)
    
        assert skycoord_inaf.separation(expected) < tolsep
        # Check we can also transform back (regression test for gh-11254).
>       roundtrip = skycoord_inaf.transform_to(ICRS())

../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/tests/test_skyoffset_transformations.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/sky_coordinate.py:627: in transform_to
    return self.__class__(new_coord, **frame_kwargs)
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/sky_coordinate.py:323: in __init__
    setattr(self, attr, skycoord_kwargs[attr])
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/sky_coordinate.py:811: in __setattr__
    frame_transform_graph.frame_attributes[attr].__get__(self)
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/attributes.py:104: in __get__
    out, converted = self.convert_input(out)
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/attributes.py:451: in convert_input
    transformedobj = value.transform_to(self._frame)
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/sky_coordinate.py:619: in transform_to
    new_coord = trans(self.frame, generic_frame)
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/transformations.py:1422: in __call__
    curr_coord = t(curr_coord, curr_toframe)
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/transformations.py:1025: in __call__
    return supcall(fromcoord, toframe)
../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/builtin_frames/icrs_observed_transforms.py:29: in icrs_to_altaz
    astrom = erfa_astrom.get().apco(altaz_frame)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

frame_or_coord = <AltAz Frame (obstime=None, location=None, pressure=0.0 hPa, temperature=0.0 deg_C, relative_humidity=0.0, obswl=1.0 micron)>

    @staticmethod
    def apco(frame_or_coord):
        '''
        Wrapper for ``erfa.apco``, used in conversions AltAz <-> ICRS and CIRS <-> ICRS
    
        Arguments
        ---------
        frame_or_coord: ``astropy.coordinates.BaseCoordinateFrame`` or ``astropy.coordinates.SkyCoord``
            Frame or coordinate instance in the corresponding frame
            for which to calculate the calculate the astrom values.
            For this function, an AltAz or CIRS frame is expected.
        '''
>       lon, lat, height = frame_or_coord.location.to_geodetic('WGS84')
E       AttributeError: 'NoneType' object has no attribute 'to_geodetic'

../../.tox/py38-test-double/lib/python3.8/site-packages/astropy/coordinates/erfa_astrom.py:45: AttributeError
```

--- a/astropy/coordinates/sky_coordinate.py
+++ b/astropy/coordinates/sky_coordinate.py
@@ -663,6 +663,13 @@ class SkyCoord(ShapedLikeNDArray):
                      set(frame_kwargs.keys())):
             frame_kwargs.pop(attr)
 
+        # Always remove the origin frame attribute, as that attribute only makes
+        # sense with a SkyOffsetFrame (in which case it will be stored on the frame).
+        # See gh-11277.
+        # TODO: Should it be a property of the frame attribute that it can
+        # or cannot be stored on a SkyCoord?
+        frame_kwargs.pop('origin', None)
+
         return self.__class__(new_coord, **frame_kwargs)
 
     def apply_space_motion(self, new_obstime=None, dt=None):

--- a/astropy/coordinates/sky_coordinate_parsers.py
+++ b/astropy/coordinates/sky_coordinate_parsers.py
@@ -437,7 +437,7 @@ def _parse_coordinate_arg(coords, frame, units, init_kwargs):
         for attr in frame_transform_graph.frame_attributes:
             value = getattr(coords, attr, None)
             use_value = (isinstance(coords, SkyCoord) or
-                         attr not in coords._attr_names_with_defaults)
+                         attr not in coords.get_frame_attr_names())
             if use_value and value is not None:
                 skycoord_kwargs[attr] = value
 

--- a/astropy/coordinates/tests/test_skyoffset_transformations.py
+++ b/astropy/coordinates/tests/test_skyoffset_transformations.py
@@ -12,6 +12,32 @@ from astropy.time import Time
 from astropy.tests.helper import assert_quantity_allclose as assert_allclose
 
 
+def test_altaz_attribute_transforms():
+    """Test transforms between AltAz frames with different attributes."""
+    el1 = EarthLocation(0*u.deg, 0*u.deg, 0*u.m)
+    origin1 = AltAz(0 * u.deg, 0*u.deg, obstime=Time("2000-01-01T12:00:00"),
+                    location=el1)
+    frame1 = SkyOffsetFrame(origin=origin1)
+    coo1 = SkyCoord(1 * u.deg, 1 * u.deg, frame=frame1)
+
+    el2 = EarthLocation(0*u.deg, 0*u.deg, 0*u.m)
+    origin2 = AltAz(0 * u.deg, 0*u.deg, obstime=Time("2000-01-01T11:00:00"),
+                    location=el2)
+    frame2 = SkyOffsetFrame(origin=origin2)
+    coo2 = coo1.transform_to(frame2)
+    coo2_expected = [1.22522446, 0.70624298] * u.deg
+    assert_allclose([coo2.lon.wrap_at(180*u.deg), coo2.lat],
+                    coo2_expected, atol=convert_precision)
+
+    el3 = EarthLocation(0*u.deg, 90*u.deg, 0*u.m)
+    origin3 = AltAz(0 * u.deg, 90*u.deg, obstime=Time("2000-01-01T12:00:00"),
+                    location=el3)
+    frame3 = SkyOffsetFrame(origin=origin3)
+    coo3 = coo2.transform_to(frame3)
+    assert_allclose([coo3.lon.wrap_at(180*u.deg), coo3.lat],
+                    [1*u.deg, 1*u.deg], atol=convert_precision)
+
+
 @pytest.mark.parametrize("inradec,expectedlatlon, tolsep", [
     ((45, 45)*u.deg, (0, 0)*u.deg, .001*u.arcsec),
     ((45, 0)*u.deg, (0, -45)*u.deg, .001*u.arcsec),
@@ -215,32 +241,6 @@ def test_m31_coord_transforms(fromsys, tosys, fromcoo, tocoo):
     roundtrip_pos = target_pos.transform_to(from_pos)
     assert_allclose([roundtrip_pos.lon.wrap_at(180*u.deg), roundtrip_pos.lat],
                     [1.0*u.deg, 1.0*u.deg], atol=convert_precision)
-
-
-def test_altaz_attribute_transforms():
-    """Test transforms between AltAz frames with different attributes."""
-    el1 = EarthLocation(0*u.deg, 0*u.deg, 0*u.m)
-    origin1 = AltAz(0 * u.deg, 0*u.deg, obstime=Time("2000-01-01T12:00:00"),
-                    location=el1)
-    frame1 = SkyOffsetFrame(origin=origin1)
-    coo1 = SkyCoord(1 * u.deg, 1 * u.deg, frame=frame1)
-
-    el2 = EarthLocation(0*u.deg, 0*u.deg, 0*u.m)
-    origin2 = AltAz(0 * u.deg, 0*u.deg, obstime=Time("2000-01-01T11:00:00"),
-                    location=el2)
-    frame2 = SkyOffsetFrame(origin=origin2)
-    coo2 = coo1.transform_to(frame2)
-    coo2_expected = [1.22522446, 0.70624298] * u.deg
-    assert_allclose([coo2.lon.wrap_at(180*u.deg), coo2.lat],
-                    coo2_expected, atol=convert_precision)
-
-    el3 = EarthLocation(0*u.deg, 90*u.deg, 0*u.m)
-    origin3 = AltAz(0 * u.deg, 90*u.deg, obstime=Time("2000-01-01T12:00:00"),
-                    location=el3)
-    frame3 = SkyOffsetFrame(origin=origin3)
-    coo3 = coo2.transform_to(frame3)
-    assert_allclose([coo3.lon.wrap_at(180*u.deg), coo3.lat],
-                    [1*u.deg, 1*u.deg], atol=convert_precision)
 
 
 @pytest.mark.parametrize("rotation, expectedlatlon", [

--- a/astropy/coordinates/tests/test_skyoffset_transformations.py
+++ b/astropy/coordinates/tests/test_skyoffset_transformations.py
@@ -56,15 +56,8 @@ def test_skyoffset(inradec, expectedlatlon, tolsep, originradec=(45, 45)*u.deg):
 
     assert skycoord_inaf.separation(expected) < tolsep
     # Check we can also transform back (regression test for gh-11254).
-    try:
-        roundtrip = skycoord_inaf.transform_to(ICRS())
-    except AttributeError as err:
-        if 'to_geodetic' in str(err):
-            pytest.xfail('See Issue 11277')
-        else:
-            raise
-    else:
-        assert roundtrip.separation(skycoord) < 1*u.uas
+    roundtrip = skycoord_inaf.transform_to(ICRS())
+    assert roundtrip.separation(skycoord) < 1*u.uas
 
 
 def test_skyoffset_functional_ra():

--- a/astropy/coordinates/tests/test_skyoffset_transformations.py
+++ b/astropy/coordinates/tests/test_skyoffset_transformations.py
@@ -335,3 +335,27 @@ def test_skyoffset_velocity_rotation(rotation, expectedpmlonlat):
     c_skyoffset0 = sc.transform_to(sc.skyoffset_frame(rotation=rotation))
     assert_allclose(c_skyoffset0.pm_lon_coslat, expectedpmlonlat[0])
     assert_allclose(c_skyoffset0.pm_lat, expectedpmlonlat[1])
+
+
+def test_skyoffset_two_frames_interfering():
+    """Regression test for gh-11277, where it turned out that the
+    origin argument validation from one SkyOffsetFrame could interfere
+    with that of another.
+
+    Note that this example brought out a different bug than that at the
+    top of gh-11277, viz., that an attempt was made to set origin on a SkyCoord
+    when it should just be stay as part of the SkyOffsetFrame.
+    """
+    # Example adapted from @bmerry's minimal example at
+    # https://github.com/astropy/astropy/issues/11277#issuecomment-825492335
+    altaz_frame = AltAz(obstime=Time('2020-04-22T13:00:00Z'),
+                        location=EarthLocation(18, -30))
+    target = SkyCoord(alt=70*u.deg, az=150*u.deg, frame=altaz_frame)
+    dirs_altaz_offset = SkyCoord(lon=[-0.02, 0.01, 0.0, 0.0, 0.0] * u.rad,
+                                 lat=[0.0, 0.2, 0.0, -0.3, 0.1] * u.rad,
+                                 frame=target.skyoffset_frame())
+    dirs_altaz = dirs_altaz_offset.transform_to(altaz_frame)
+    dirs_icrs = dirs_altaz.transform_to(ICRS())
+    target_icrs = target.transform_to(ICRS())
+    # The line below was almost guaranteed to fail.
+    dirs_icrs.transform_to(target_icrs.skyoffset_frame())

--- a/docs/changes/coordinates/11730.bugfix.rst
+++ b/docs/changes/coordinates/11730.bugfix.rst
@@ -1,0 +1,5 @@
+Different ``SkyOffsetFrame`` classes no longer interfere with each other,
+causing difficult to debug problems with the ``origin`` attribute. The
+``origin`` attribute now no longer is propagated, so while it remains
+available on a ``SkyCoord`` that is an offset, it no longer is available once
+that coordinate is transformed to another frame. [#11730]

